ambry-replication/src/main/java/com.github.ambry.replication/ReplicationManager.java

```diff
@@ -595,33 +595,36 @@ private void readFromFileAndPersistIfNecessary(String mountPath) throws Replicat
               FindToken token = factory.getFindToken(stream);
               // update token
               PartitionInfo partitionInfo = partitionsToReplicate.get(partitionId);
-              boolean updatedToken = false;
-              for (RemoteReplicaInfo remoteReplicaInfo : partitionInfo.getRemoteReplicaInfos()) {
-                if (remoteReplicaInfo.getReplicaId().getDataNodeId().getHostname().equalsIgnoreCase(hostname)
-                    && remoteReplicaInfo.getReplicaId().getDataNodeId().getPort() == port
-                    && remoteReplicaInfo.getReplicaId().getReplicaPath().equals(replicaPath)) {
-                  logger.info("Read token for partition {} remote host {} port {} token {}", partitionId, hostname,
-                      port, token);
-                  if (partitionInfo.getStore().getSizeInBytes() > 0) {
-                    remoteReplicaInfo.initializeTokens(token);
-                    remoteReplicaInfo.setTotalBytesReadFromLocalStore(totalBytesReadFromLocalStore);
-                  } else {
-                    // if the local replica is empty, it could have been newly created. In this case, the offset in
-                    // every peer replica which the local replica lags from should be set to 0, so that the local
-                    // replica starts fetching from the beginning of the peer. The totalBytes the peer read from the
-                    // local replica should also be set to 0. During initialization these values are already set to 0,
-                    // so we let them be.
-                    tokenWasReset = true;
-                    replicationMetrics.replicationTokenResetCount.inc();
-                    logger.info("Resetting token for partition {} remote host {} port {}, persisted token {}",
-                        partitionId, hostname, port, token);
+              if (partitionInfo != null) {
```

Comment:
Okay, I see what you mean about it being reset anyway, since the persistor.write() call during shutdown will ignore all of the stores that did not start up. Since this is the case, I think the tokens should be reset during initialization, to ensure that the token reset is also persisted when a clean shutdown does not occur.